Add Host Reports widget and Host Reports status chart widget

adiabramovitch opened the pull request.

lzap commented:
Before we start, a recap of what we have in core today. This is `app/views/dashboard/_reports_widget.html.erb`:

![Snímek obrazovky 2022-01-27 v 12 26 23](https://user-images.githubusercontent.com/49752/151350030-a3694147-32d8-4a75-9320-39261542bd84.png)

The following is `app/views/dashboard/_status_widget.html.erb` which is evaluated for each individual origin/type:

![Snímek obrazovky 2022-01-27 v 12 27 31](https://user-images.githubusercontent.com/49752/151350206-ed5c0a9b-f456-4088-8baa-ab2ac6d45ea5.png)

The same data but graphically is presented in `app/views/dashboard/_status_chart_widget.html.erb` again for each type:

![Snímek obrazovky 2022-01-27 v 12 28 45](https://user-images.githubusercontent.com/49752/151350394-25301a55-d9dc-42ab-a8f9-8f3bd523d4b0.png)

And finally we have distribution charts which we will not carry over to the plugin, at least for the first version. Defined in `app/views/dashboard/_distribution_widget.html.erb`:

![Snímek obrazovky 2022-01-27 v 12 29 49](https://user-images.githubusercontent.com/49752/151350562-f10060cf-9d18-476f-b1b1-a3fdb892feab.png)

lzap commented on app/views/dashboard/_host_reports_widget.html.erb:
```diff
@@ -0,0 +1,27 @@
+<h4 class="header">
+  <%= _("Host Latest Reports") %>
```
I am thinking like "Latest Host Reports".

lzap commented on app/views/dashboard/_host_reports_widget.html.erb:
```diff
@@ -0,0 +1,27 @@
+        </tr>
+      <% end %>
+    </tbody>
+  </table>
```
This table will look much more simple: host, change, nochange, failure. Headers would be: Host, Changed, Unchanged, Failed.

lzap commented on app/views/dashboard/_host_reports_widget.html.erb:
```diff
@@ -0,0 +1,27 @@
+  <table class="<%= table_css_classes 'table-fixed reports-table' %>">
+    <thead>
+      <tr>
+        <%= latest_headers %>
```
Do not use this helper I think, this is so simple table that you can simply do ERB.

lzap commented:
I think keeping "for all" variants does not make sense. How should we handle systems that has only one or another? Should we report the missing report as a failure? How would users interpret this? @ares

lzap commented on app/views/dashboard/_host_reports_status_links.html.erb:
```diff
@@ -0,0 +1,4 @@
+
+<%= searchable_links _('Hosts Reports that had changed'), "change > 0", :active_hosts_ok_enabled %>
+<%= searchable_links _('Hosts Reports that failed'), "failure > 0" , :bad_hosts_enabled %>
+<%= searchable_links _('Hosts Reports that have not changed'), "nochange > 0", :reports_missing %>
```
* Host Reports with changes
* Host Reports without changes
* Host Reports with failures

Perhaps?

lzap replied:
Sorry correction:

* Hosts with changes
* Hosts without changes
* Hosts with failures
* Hosts with disabled alerts

lzap commented on app/views/dashboard/_host_reports_status_widget.html.erb:
```diff
@@ -0,0 +1,11 @@
+  <% if @data.settings[:format] %>
+    <%= _('Host Reports Status for %s') % @data.settings[:format] %>
+  <% else %>
+    <%= _('Host Reports Status') %>
```
Puppet Host Reports
Ansible Host Reports
All Host Reports

Perhaps?

lzap commented on app/views/dashboard/_host_reports_status_chart_widget.html.erb:
```diff
@@ -0,0 +1,15 @@
+    <%= title = _('Host Chart for %s') % format %>
+  <% else %>
+    <%= title = _('Host Chart') %>
+  <% end %>
```
Puppet Host Reports Chart
Ansible ...
All ...

Perhaps?

lzap commented on app/views/dashboard/_host_reports_widget.html.erb:
```diff
@@ -0,0 +1,27 @@
+<h4 class="header">
+  <%= _("Host Latest Reports") %>
+</h4>
+<% if @data.latest_events? %>
```
In plugin widgets you cannot load data into `@data` in app/services/dashboard/data.rb you have to do it in the ERB template. You can share the SQL queries as scopes or methods in either HostReport or Host::Managed models (the plugin has a concern class for extensions). Example: https://github.com/theforeman/foreman_discovery/blob/develop/app/views/dashboard/_discovery_widget.html.erb#L1

lzap commented on lib/foreman_host_reports/engine.rb:
```diff
@@ -40,6 +40,17 @@ class Engine < ::Rails::Engine
+          HostReport.formats.each do |format, _|
+            if format != 'plain' && (format = format.capitalize)
+              widget 'host_reports_status_chart_widget', :name => N_("Host Reports Status Chart for %s") % format, :sizex => 4, :sizey => 1, settings: { format: format }
+              widget 'host_reports_status_widget', :name => N_("Host Reports Status for %s") % format, :sizex => 8, :sizey => 1, settings: { format: format }
```
[FORMAT] Host Reports Chart
[FORMAT] Host Reports

Feels better to me.

lzap reviewed (commented):
Looks ok, now the most important part - the data. As I said in one of the comments, you need to construct queries which will be completely different - they don't exist yet. You will probably need to use joins, groups, counts so it would be useful to create them as either scopes or methods in model classes so in ERB you can simply call them easily.

Let me know if you need help.

lzap commented:
Warning: For "Last Host Reports" table/widget you will need to search hosts which had new reports uploaded recently. If you read the code in core, its using "recent" scope which use "with_last_report_within". This needs to have "last_report" field set, this has been only implemented recently:

https://github.com/theforeman/foreman_host_reports/pull/10/files#diff-740ba7fd836ee4283ba29c67cb4b50c242119b03f454d68569fd3fd5e695be35R83

If you can review this patch and also test it, that would be awesome.

One remark for the "[FORMAT] Host Reports" table, I forgot one important line we should print (4th line): **Hosts with disabled alerts**. The query is easy: `hosts.alerts_disabled.count` that is unchanged from the old reports.

lzap commented:
Data requirements:

Latest Host Reports: list of hosts ordered by last_report limited by 10 (or whatever fills into the widget) available in the context (Organization, Location, filters for the signed-in user)

Puppet/Ansible/All Host Reports:

Hosts with changes: total count of hosts which had in their last report change > 0 for the user's context
Hosts without changes: total count of hosts which had in their last report nochange > 0 for the user's context
Hosts with failures: total count of hosts which had in their last report failure > 0 for the user's context
Hosts with disabled alerts: total count of hosts with enabled = false for the user's context
Total hosts: total count of all hosts for the user's context

@ares can you confirm perhaps my understanding of these results? the original code in core is, ehm, unreadable I would say :)

lzap commented on app/models/host_report.rb:
```diff
@@ -55,4 +55,53 @@ def self.search_by_keyword(_key, operator, value)
+  end
+
+  def self.report_changes
+    hosts.search_for("report_changes = 1")
```
If you can avoid scoped search, avoid it. It creates ugly SQL queries. I think in this case this could be a simple join.

Also report_changes is an alias to host_report.change which is an integer. So you are searching for `change > 0` not `= 1`. The same below.

lzap replied:
Still TODO.

lzap commented on app/models/host_report.rb:
```diff
@@ -55,4 +55,53 @@ def self.search_by_keyword(_key, operator, value)
+      :disabled_hosts => disabled_alerts.count,
+      :reports_missing => report_missing.count
+    )
+  end
```
Probably try to find a better method name - this is HostReport model and it is weird. Perhaps `totals` or `counts` or something like that. The helper methods should be `private`.

lzap replied:
My comment still stands.

lzap commented on app/models/host_report.rb:
```diff
@@ -55,4 +55,53 @@ def self.search_by_keyword(_key, operator, value)
+  def self.hosts
+    filter = nil # TODO: implement global filtering
+    hosts || Host.authorized(:view_hosts, Host).search_for(filter)
+  end
```
What this method is supposed to return? I don't think it should be in this model, host report has exactly one host. Just merge it with the `latest_events` method it is short enough.

Ron-Lavi replied:
I am getting an endless loop from line `97` for some reason,
see: https://pastebin.com/bG6Hww6t

adiabramovitch replied:
@Ron-Lavi Fixed it, I'll push is soon

adiabramovitch replied:
I think the hosts function is needed. 
We are searching through the relevant Hosts (with the right permission and filter) which hosts are corresponded to the scoped search - "report_failures", "report_changes" and so on..
correct me if I am wrong, without that function I'll need to write these lines at every helper method

lzap replied:
What I was trying to say is this method alone will not use much use, so I would simply move/merge it with the `latest_events`.

lzap commented on app/models/host_report.rb:
```diff
@@ -55,4 +55,53 @@ def self.search_by_keyword(_key, operator, value)
+                                 .reorder(:last_report => :desc)
+                                 .limit(9)
+                                 .preload(:host)
+  end
```
Please rename, let's kill the old term "event".

adiabramovitch replied:
lateset_reports are OK?

lzap replied:
Yup that would work for me.

lzap reviewed (changes requested):
Good direction, couple of comments.

Now, most importantly, this needs unit tests. Please utilize the FactoryBot we have and create various test cases and test every single usecase that can happen. Disabled hosts, missing reports, no reports, errors, no errors, changes and all the results that the new `total` method returns.

lzap commented:
Accidentally committed `.idea/workspace.xml` feel free to add `/idea` to the `.gitignore`.

adiabramovitch committed:
Add Host Reports widgets

adiabramovitch commented:
So few things - 
1. The tests are just a draft, didn't have the chance to update the Factories
2. Now the data for the widgets is the same, I am not differentiating between Puppet and Ansible. I need to find a smart way to pass the format to the helper functions without passing it to each function. 
will be happy to hear any idea, I thought about an instance variable, not sure if it is the right approach or there is some rails solution I am not aware of. 
3. Not Sure which colors do you want to use for the widgets if there is any p[prefrence
4. The alignment of the "__ Host Reports" is off, if @Ron-Lavi  you can take a look, please.

Current Status: 
![image](https://user-images.githubusercontent.com/57755873/152950840-b9461cfd-dade-40b4-bdb7-26d963228410.png)

lzap commented on app/helpers/foreman_host_reports/reports_dashboard_helper.rb:
```diff
@@ -0,0 +1,92 @@
+  module ReportsDashboardHelper
+
+    def searchable_links(name, search, counter)
+      # TODO: implement global filtering
```
Please drop this comment, it looks like in the thread I opened on Discourse people tend to agree that global search can go away.

lzap commented on app/helpers/foreman_host_reports/reports_dashboard_helper.rb:
```diff
@@ -0,0 +1,92 @@
+        :failure_hosts => report_failures.count,
+        :disabled_hosts => disabled_alerts.count
+      )
+    end
```
Nice! I think this is much better place than the model.

lzap replied:
Memorization of a hash can be done in a nicer way I think:

    @report_totals ||= { ... }

lzap replied:
Correction: you need to call `totals` per each report type, therefore you cannot memorize it this way. Store the data in a hash indexed by type then.

lzap replied:
@ezr-ondrej or @ares can you give me advice here, I was researching this a bit but I am unsure.

My concern is thread safety of member variables in view/controller helpers. This probably stems from my lack of understanding how module including/extending/prepending works in Ruby. Do multiple RoR requests share the helper module therefore this memorization would not be safe? Or does RoR creates new instance of a helper per thread so it's actually safe.

See, widgets don't have plugin API that would allow defining data sources, therefore, we must load them either in the view itself or (probably better) in a helper. Thanks.

ezr-ondrej replied:
Templates are thread safe, but there is worse issue with this. This model is included in a global namespace of view, so the name `totals` can be quite ambigous and another helper having the same method with a memoization would cause total chaos, that would be very hard to debug as the method would return your data, or your method would return it's data, that is why it is better to avoid memoization in a helper, if really necessary, then please call it less ambigous, but I'd rather if we avoid it entirely.

I agree that there is not very good API to get custom data, but it has the `Dashboard::Data` class, that you get instance of while rendering you widget. I believe it could be used to put these methods in. Given that current `reports_widget` is using it, we might even want to override [`latest_events`](https://github.com/theforeman/foreman/blob/fdd6b3cd41e58098f0da2fab828c42c4de3d8e10/app/services/dashboard/data.rb#L27) and just feed it our data?

If you don't like overriding and patching that class, we could just have our own `Dashboard::HostReportsData` and patch dashboard controller to add aditional `before_action` and load this our custom data class there similar to what the [controller does now to load the data](https://github.com/theforeman/foreman/blob/fdd6b3cd41e58098f0da2fab828c42c4de3d8e10/app/controllers/dashboard_controller.rb#L82)

lzap commented on app/helpers/foreman_host_reports/reports_dashboard_helper.rb:
```diff
@@ -0,0 +1,92 @@
+    def hosts
+      filter = nil # TODO: implement global filtering
+      Host.authorized(:view_hosts, Host).search_for(filter)
+    end
```
If this is used only in that one method, move it there. Drop the comment please.

lzap commented on app/helpers/foreman_host_reports/reports_dashboard_helper.rb:
```diff
@@ -0,0 +1,92 @@
+    end
+
+    def report_changes
+      Host.joins(:host_reports).where("change > 0 ").uniq
```
There is an extra space.

lzap commented on app/helpers/foreman_host_reports/reports_dashboard_helper.rb:
```diff
@@ -0,0 +1,92 @@
+
+    def report_failures
+      Host.joins(:host_reports).where("failure > 0").uniq
+    end
```
Can you make all the four methods actually to return count? That is safer, we never want to load all the records into memory so let's not even return it.

lzap commented on app/models/host_report.rb:
```diff
@@ -55,4 +55,4 @@ def self.search_by_keyword(_key, operator, value)
     }
   end
-end
+end
```
Keep endlines if possible, that's the preferred style.

lzap commented on test/unit/dashboard_test.rb:
```diff
@@ -0,0 +1,36 @@
+  end
+
+  test 'check Hosts with disabled alerts ' do
+    host1.update(:enabled => false)
```
I would suggest you update the factory bot with some traits that can be shared across tests. We will likely get many more tests in the future.

lzap reviewed (changes requested):
The code looks good. Few nitpicks.

But it does not work for me:

```
11:10:56 rails.1   | 2022-02-08T11:10:56 [I|app|602d4a95] Backtrace for 'Failed to load widget' error (ActionView::Template::Error): no implicit conversion of nil into String
11:10:56 rails.1   |  602d4a95 | /home/lzap/work/foreman_host_reports/app/helpers/foreman_host_reports/reports_dashboard_helper.rb:7:in `+'
11:10:56 rails.1   |  602d4a95 | /home/lzap/work/foreman_host_reports/app/helpers/foreman_host_reports/reports_dashboard_helper.rb:7:in `block in searchable_links'
11:10:56 rails.1   |  602d4a95 | /home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/actionview-6.0.3.7/lib/action_view/helpers/capture_helper.rb:45:in `block in capture'
11:10:56 rails.1   |  602d4a95 | /home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/actionview-6.0.3.7/lib/action_view/helpers/capture_helper.rb:209:in `with_output_buffer'
11:10:56 rails.1   |  602d4a95 | /home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/actionview-6.0.3.7/lib/action_view/helpers/capture_helper.rb:45:in `capture'
11:10:56 rails.1   |  602d4a95 | /home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/actionview-6.0.3.7/lib/action_view/helpers/tag_helper.rb:273:in `content_tag'
11:10:56 rails.1   |  602d4a95 | /home/lzap/work/foreman_host_reports/app/helpers/foreman_host_reports/reports_dashboard_helper.rb:6:in `searchable_links'
11:10:56 rails.1   |  602d4a95 | /home/lzap/work/foreman/app/views/dashboard/_status_links.html.erb:6:in `_app_views_dashboard__status_links_html_erb___2276221108910539187_216060'
11:10:56 rails.1   |  602d4a95 | /home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/actionview-6.0.3.7/lib/action_view/base.rb:274:in `_run'
11:10:56 rails.1   |  602d4a95 | /home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/actionview-6.0.3.7/lib/action_view/template.rb:185:in `block in render'
11:10:56 rails.1   |  602d4a95 | /home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/activesupport-6.0.3.7/lib/active_support/notifications.rb:182:in `instrument'
11:10:56 rails.1   |  602d4a95 | /home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/actionview-6.0.3.7/lib/action_view/template.rb:385:in `instrument_render_template'
11:10:56 rails.1   |  602d4a95 | /home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/actionview-6.0.3.7/lib/action_view/template.rb:183:in `render'
11:10:56 rails.1   |  602d4a95 | /home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/actionview-6.0.3.7/lib/action_view/renderer/partial_renderer.rb:357:in `
```

lzap commented:
> Now the data for the widgets is the same, I am not differentiating between Puppet and Ansible. I need to find a smart way to pass the format to the helper functions without passing it to each function.
will be happy to hear any idea, I thought about an instance variable, not sure if it is the right approach or there is some rails solution I am not aware of.

Not sure what you mean.

> Not Sure which colors do you want to use for the widgets if there is any p[prefrence

The same colours as we will end up on the index page I guess? Or you could use the very same icons too.

lzap commented:
@adiabramovitch can you break the big `totals` method into smaller methods and just use them? Also we talked about implementing the `format` method argument. I don't see this in the last version.

adiabramovitch commented on app/helpers/foreman_host_reports/reports_dashboard_helper.rb:
```diff
@@ -0,0 +1,88 @@
+        content_tag(:span, raw('&nbsp;'), :class => 'label', :style => "background-color:" + host_reports_report_color[counter]) +
+          raw('&nbsp;') +
+          link_to(name, hosts_path(:search => search), :class => "dashboard-links") +
+          content_tag(:h4, totals(format)[counter])
```
@lzap Because of the use of totals here, I kept the function but I changed the usage in the html.erb files. 
Not sure what did you suggested regarding the format argument, it is required here and in the other helper functions.

lzap replied:
I see, well this ERB is correct, however, I will drop a comment below.

lzap commented on app/helpers/foreman_host_reports/reports_dashboard_helper.rb:
```diff
@@ -0,0 +1,88 @@
+        :failure_hosts => report_failures(format),
+        :disabled_hosts => disabled_alerts,
+      }
+    end
```
We cannot memorize values with arguments, each call should yield to a different result. I suggest to drop all memorization for now, from all methods in this helper.

lzap commented:
So I deleted all my reports and loaded all fixtures from the repo, I have the following dataset (all reports do belong to a single host):

```
HostReport.all.pluck(:id, :format, :change, :nochange, :failure)
=> [[342, "ansible", 0, 2, 0],
 [341, "ansible", 0, 2, 0],
 [340, "ansible", 0, 2, 0],
 [339, "ansible", 0, 2, 0],
 [338, "ansible", 0, 2, 0],
 [337, "ansible", 0, 2, 0],
 [336, "ansible", 0, 2, 0],
 [335, "ansible", 0, 2, 0],
 [334, "ansible", 1, 0, 0],
 [333, "ansible", 2, 0, 0],
 [332, "ansible", 2, 0, 0],
 [331, "ansible", 2, 0, 0],
 [330, "ansible", 2, 0, 0],
 [323, "ansible", 0, 1, 0],
 [324, "ansible", 1, 0, 0],
 [325, "ansible", 0, 1, 0],
 [326, "ansible", 0, 1, 1],
 [327, "ansible", 0, 6, 0],
 [328, "ansible", 0, 6, 0],
 [329, "ansible", 1, 0, 0],
 [315, "puppet", 0, 7, 0],
 [314, "ansible", 0, 1, 0],
 [316, "puppet", 1, 7, 0],
 [317, "puppet", 0, 8, 0],
 [318, "puppet", 0, 7, 1],
 [319, "ansible", 0, 7, 0],
 [320, "ansible", 0, 7, 0],
 [321, "ansible", 0, 1, 0],
 [322, "ansible", 1, 0, 0]]
```

The last_reports widgets show a SQL error:

![Snímek obrazovky 2022-02-16 v 15 51 28](https://user-images.githubusercontent.com/49752/154299883-2b09d5a8-f59b-4f02-9328-d44b591b1ea3.png)

Other two widgets have numbers off, I would expect them to be on the right side, is this intentional?

![Snímek obrazovky 2022-02-16 v 16 37 54](https://user-images.githubusercontent.com/49752/154299980-d91bcf03-5295-4881-9321-cb940255900c.png)

But there is one problem - we need to count only last reports, the query appears to count across all reports per host. What we want is to count hosts for which their last reports have failures/unchanged/changed.

Rubocop appears to have issues.

ezr-ondrej commented on app/helpers/foreman_host_reports/reports_dashboard_helper.rb:
```diff
@@ -0,0 +1,70 @@
+
+    def latest_reports
+      hosts = Host.authorized(:view_hosts, Host)
+      @latest_reports = HostReport.authorized(:view_host_reports).my_reports
```
Setting an instance value in a helper is not great, because you might override value from controller and it is hard to track down where the value gets overriden (especially when the code that gets called is decided by what widgets you have on the dashboard).
Do we need to set the variable here?

adiabramovitch replied:
In that case, it is redundant. 
Thank you

adiabramovitch commented:
> So I deleted all my reports and loaded all fixtures from the repo, I have the following dataset (all reports do belong to a single host):
> 
> ```
> HostReport.all.pluck(:id, :format, :change, :nochange, :failure)
> => [[342, "ansible", 0, 2, 0],
>  [341, "ansible", 0, 2, 0],
>  [340, "ansible", 0, 2, 0],
>  [339, "ansible", 0, 2, 0],
>  [338, "ansible", 0, 2, 0],
>  [337, "ansible", 0, 2, 0],
>  [336, "ansible", 0, 2, 0],
>  [335, "ansible", 0, 2, 0],
>  [334, "ansible", 1, 0, 0],
>  [333, "ansible", 2, 0, 0],
>  [332, "ansible", 2, 0, 0],
>  [331, "ansible", 2, 0, 0],
>  [330, "ansible", 2, 0, 0],
>  [323, "ansible", 0, 1, 0],
>  [324, "ansible", 1, 0, 0],
>  [325, "ansible", 0, 1, 0],
>  [326, "ansible", 0, 1, 1],
>  [327, "ansible", 0, 6, 0],
>  [328, "ansible", 0, 6, 0],
>  [329, "ansible", 1, 0, 0],
>  [315, "puppet", 0, 7, 0],
>  [314, "ansible", 0, 1, 0],
>  [316, "puppet", 1, 7, 0],
>  [317, "puppet", 0, 8, 0],
>  [318, "puppet", 0, 7, 1],
>  [319, "ansible", 0, 7, 0],
>  [320, "ansible", 0, 7, 0],
>  [321, "ansible", 0, 1, 0],
>  [322, "ansible", 1, 0, 0]]
> ```
> 
> The last_reports widgets show a SQL error:
> 
> ![Snímek obrazovky 2022-02-16 v 15 51 28](https://user-images.githubusercontent.com/49752/154299883-2b09d5a8-f59b-4f02-9328-d44b591b1ea3.png)
> 


Not sure why you have this error because I deleted my reports as well: 
```
[[1614, 0, 6, 0, "ansible"],
 [1615, 0, 1, 1, "ansible"],
 [1616, 0, 1, 0, "ansible"],
 [1617, 1, 0, 0, "ansible"],
 [1618, 0, 1, 0, "ansible"],
 [1619, 1, 0, 0, "ansible"],
 [1620, 0, 1, 0, "ansible"],
 [1621, 0, 7, 0, "ansible"],
 [1622, 0, 7, 0, "ansible"],
 [1623, 0, 1, 0, "ansible"],
 [1624, 1, 0, 0, "ansible"],
 [1625, 0, 6, 0, "ansible"],
 [1626, 0, 6, 0, "ansible"],
 [1586, 0, 7, 0, "ansible"],
 [1587, 0, 1, 0, "ansible"],
 [1588, 1, 0, 0, "ansible"],
 [1589, 0, 6, 0, "ansible"],
 [1590, 0, 6, 0, "ansible"],
 [1591, 0, 1, 1, "ansible"],
 [1592, 0, 1, 0, "ansible"],
 [1593, 1, 0, 0, "ansible"],
 [1594, 0, 1, 0, "ansible"],
 [1595, 1, 0, 0, "ansible"],
 [1596, 0, 1, 0, "ansible"],
 [1597, 0, 7, 0, "ansible"],
 [1598, 0, 7, 0, "ansible"],
 [1599, 0, 1, 0, "ansible"],
 [1600, 1, 0, 0, "ansible"],
 [1601, 0, 6, 0, "ansible"],
 [1602, 0, 6, 0, "ansible"],
 [1603, 0, 1, 1, "ansible"],
 [1604, 0, 1, 0, "ansible"],
```
and I see the last_reports widget:
![image](https://user-images.githubusercontent.com/57755873/154442187-a4982fd0-638c-400b-b97f-99751081408a.png)

> Other two widgets have numbers off, I would expect them to be on the right side, is this intentional?

About the numbers, I noticed it too but I haven't changed it from the core dashboard so @Ron-Lavi can you take a look, please?
> 
> ![Snímek obrazovky 2022-02-16 v 16 37 54](https://user-images.githubusercontent.com/49752/154299980-d91bcf03-5295-4881-9321-cb940255900c.png)
> 
> But there is one problem - we need to count only the last reports, the query appears to count across all reports per host. What we want is to count hosts for which their last reports have failures/unchanged/changed.

Regarding the query, we want to make a match between the last_report of the host to the report itself in the HostReport table but from my understanding, we don't have the id of that report just its datetime. a property we can't do a  1:1  match. 
please correct me if I am wrong or if I have missed something.

> 
> Rubocop appears to have issues.

Fixed the rubocop issues.

Ron-Lavi commented on lib/foreman_host_reports/engine.rb:
```diff
@@ -44,6 +44,19 @@ class Engine < ::Rails::Engine
+          HostReport.formats.each do |format, _|
+            if format != 'plain' && (format = format.capitalize)
+              widget 'host_reports_status_chart_widget', :name => N_(" %s Host Reports Chart ") % format, :sizex => 4, :sizey => 1, settings: { format: format }
+              widget 'host_reports_status_widget', :name => N_(" %s Host Reports ") % format, :sizex => 8, :sizey => 1, settings: { format: format }
```
can you change the `sizex` to 4?

Ron-Lavi commented on app/helpers/foreman_host_reports/reports_dashboard_helper.rb:
```diff
@@ -0,0 +1,70 @@
+        content_tag(:span, sanitize('&nbsp;'), :class => 'label', :style => "background-color: #{host_reports_report_color[counter]}") +
+          sanitize('&nbsp;') +
+          link_to(name, hosts_path(:search => search), :class => "dashboard-links") +
+          content_tag(:h4, send(counter, format))
```
if you can, change it to span and try to `pull-right` the count

lzap commented:
One more observation, use full column names instead of C,U,F there is a plenty of space.

I will look into the query now.

lzap commented on app/helpers/foreman_host_reports/reports_dashboard_helper.rb:
```diff
@@ -0,0 +1,74 @@
+                .where(host_id: hosts.pluck(:id))
+                .reorder(:last_report => :desc)
+                .limit(9)
+                .preload(:host)
```
I have realized that the my_reports scope actually uses a helper to join host authorized host reports:

```ruby
  scope :my_reports, lambda {
    if !User.current.admin? || Organization.expand(Organization.current).present? || Location.expand(Location.current).present?
      joins_authorized(Host, :view_hosts)
    end
  }
```

Could you simplify the query to just:

```ruby
      HostReport.authorized(:view_host_reports).my_reports
                .reorder(:last_report => :desc)
                .limit(9)
                .preload(:host)
```

removing the where condition and the helper? Then try this with hosts belonging to different organizations (A,B) and try it under regular user (non-admin) from org A. You should not see hosts from org B.

lzap commented on app/helpers/foreman_host_reports/reports_dashboard_helper.rb:
```diff
@@ -0,0 +1,74 @@
+
+    def failure_hosts(format)
+      Host.joins(:host_reports).where("format = ? and failure > 0", HostReport.formats[format]).uniq.count
+    end
```
So I spent quite some time trying to figure out a single huge query that would do all three selects via SQL list with subselect but then I ditched the idea completely it got extremely complicated once taxonomy got involved. This is a tested way how to fetch the data:

```ruby
HostReport.authorized(:view_host_reports).my_reports.where("format = ? and nochange > 0", HostReport.formats[:puppet]).reorder('').group(:host_id).maximum(:id).count
```

I tested this with taxonomy in mind too, after I created UserA that only belongs to OrgA which only has a one host:

```
User.current = User.find(6)

[34] pry(main)> HostReport.authorized(:view_host_reports).my_reports.where("format = ? and nochange > 0", HostReport.formats[:puppet]).reorder('').group(:host_id).maximum(:id).count

2022-02-21T18:08:44 [D|per|] checking permission view_host_reports for class HostReport
2022-02-21T18:08:44 [D|per|] organization_ids: []
2022-02-21T18:08:44 [D|per|] location_ids: []
2022-02-21T18:08:44 [D|per|] filter with role_id: 12 limited: false search:  taxonomy_search:
2022-02-21T18:08:44 [D|per|] checking permission view_hosts for class Host
2022-02-21T18:08:44 [D|per|] organization_ids: []
2022-02-21T18:08:44 [D|per|] location_ids: []
2022-02-21T18:08:44 [D|per|] filter with role_id: 2 limited: false search:  taxonomy_search:
 | filter with role_id: 3 limited: false search:  taxonomy_search:
 | filter with role_id: 6 limited: false search:  taxonomy_search:
 | filter with role_id: 7 limited: false search:  taxonomy_search:
 | filter with role_id: 8 limited: false search:  taxonomy_search:
 | filter with role_id: 9 limited: false search:  taxonomy_search:
 | filter with role_id: 15 limited: false search:  taxonomy_search:
 | filter with role_id: 16 limited: false search:  taxonomy_search:
 | filter with role_id: 24 limited: false search:  taxonomy_search:
 | filter with role_id: 25 limited: false search:  taxonomy_search:
 | filter with role_id: 27 limited: false search:  taxonomy_search:
=> 1
```

It correctly shows just one host while there are multiple in other orgs in my case. This is the change we need to do.

adiabramovitch replied:
I am not sure I fully understand what you did. 
The problem you mentioned is:

> But there is one problem - we need to count only the last reports, the query appears to count across all reports per host. What we want is to count hosts for which their last reports have failures/unchanged/changed.

Does this fix it? because I don't see something related to host.last_report in the query.

lzap replied:
> Does this fix it?

Yes it does, note the group and max aggregate functions.

lzap commented on app/helpers/foreman_host_reports/reports_dashboard_helper.rb:
```diff
@@ -0,0 +1,72 @@
+      HostReport.authorized(:view_host_reports).my_reports
+                .reorder(:last_report => :desc)
+                .limit(9)
+                .preload(:host)
```
This query does not work and it errors out. It renders to the following SQL: `"SELECT \"host_reports\".* FROM \"host_reports\" ORDER BY \"last_report\" DESC LIMIT 9"` (under admin account so it is more simple). The reason is that last_report must be changed to reported_at:

    HostReport.authorized(:view_host_reports).my_reports.reorder(:reported_at => :desc).limit(9).preload(:host)

I am not sure how this could work for you, last report widgets could never display anything but the SQL error.

lzap replied:
Also please add a test case for this method, it is untested at the moment.

lzap reviewed (changes requested):
Looks good except the last reports widget does not work due to typo in the SQL statement.

lzap commented on app/views/dashboard/_host_reports_status_links.html.erb:
```diff
@@ -0,0 +1,3 @@
+<%= host_reports_searchable_links _('Host Reports with changes'), "report_changes > 0", :change_hosts, format%>
+<%= host_reports_searchable_links _('Host Reports without changes'), "report_nochanges > 0", :nochange_hosts, format %>
+<%= host_reports_searchable_links _('Host Reports with failures'), "report_failures > 0" , :failure_hosts, format %>
```
Found a bug = the widget always shows summary for particular format (Ansible, Puppet) but the query does not reflect it. These queries must be "report_changes > 0 and report_format = 'Ansible'" or something like that. 

However, to be able to do that, you need to update host_extensions.rb:

```
      scoped_search relation: :host_reports, on: :change, rename: :report_changes, only_explicit: true
      scoped_search relation: :host_reports, on: :nochange, rename: :report_nochanges, only_explicit: true
      scoped_search relation: :host_reports, on: :failure, rename: :report_failures, only_explicit: true
```

Here add three more lines to add such a scoped search.

lzap reviewed (changes requested):
It works except links.

lzap commented:
Other than that it works fine, charts for some reason show different main value in the middle. Is that intentional? How does that work? I would suggest to show percent of failures by default on all:

![Snímek obrazovky 2022-03-02 v 14 24 13](https://user-images.githubusercontent.com/49752/156370308-aca669c7-a9b9-4e94-8424-f4a621fab573.png)

adiabramovitch commented:
> Other than that it works fine, charts for some reason show different main value in the middle. Is that intentional? How does that work? I would suggest to show percent of failures by default on all:
> 
> ![Snímek obrazovky 2022-03-02 v 14 24 13](https://user-images.githubusercontent.com/49752/156370308-aca669c7-a9b9-4e94-8424-f4a621fab573.png)

I am not sure, I think it is related to react_component
@Ron-Lavi  is it?

lzap reviewed (approved):
Works great, I am bit confused by the chart but it's okay. Thank you, great work! :)
